File: destack/test/language/node.py
from destack import (
    LineShape2D,
    Node,
    NodeType,
    Session,
    Shape2D,
)


def test_node_inheritance(session: Session):
    """Test the Node inheritance hierarchy."""
    assert Node.metatype == NodeType.NODE
    assert Node.__definition__.is_abstract
    assert Shape2D.metatype == NodeType.SHAPE2D
    assert Shape2D.__definition__.is_abstract
    assert LineShape2D.__definition__.base_type == Shape2D.metatype
    assert LineShape2D.__definition__.inherits == [
        NodeType.NODE,
        NodeType.ENTITY,
        NodeType.ENTITY2D,
        NodeType.SHAPE2D,
    ]
    assert len(Node.__definition__.inherited_by) == len(NodeType) - 1

File: destack-rs/destack/src/core/definition/node.rs
//! destack.core.definition.node

#![destack::partial(destack.core.definition.node, file)]

use crate::{
    ActionDefinition, ConstantDefinition, ConstraintDefinition, EncoderStability, IndexDefinition,
    MethodDefinition, NodeType, PermissionDefinition, PropertyDefinition, StructType,
    TagDefinition, TraitType, UniverseCategory, UniverseDomain,
};

#[destack::generated(NodeDefinition, -, block)]
/// Definition of a builtin Node.
pub struct NodeDefinition {
    pub id: u32,
    pub r#type: NodeType,
    pub name: String,
    pub description: String,
    pub domain: UniverseDomain,
    pub category: UniverseCategory,
    pub stability: EncoderStability,
    pub is_abstract: bool,
    pub is_final: bool,
    pub is_singleton: bool,
    pub properties: Vec<PropertyDefinition>,
    pub indexes: Vec<IndexDefinition>,
    pub constraints: Vec<ConstraintDefinition>,
    pub permissions: Vec<PermissionDefinition>,
    pub methods: Vec<MethodDefinition>,
    pub actions: Vec<ActionDefinition>,
    pub constants: Vec<ConstantDefinition>,
    pub tags: Vec<TagDefinition>,
    pub base_type: Option<NodeType>,
    pub extended_by: Vec<NodeType>,
    pub inherits: Vec<NodeType>,
    pub inherited_by: Vec<NodeType>,
    pub traits: Vec<TraitType>,
    pub self_traits: Vec<TraitType>,
    pub parent_types: Vec<NodeType>,
    pub child_types: Vec<NodeType>,
    pub ancestor_types: Vec<NodeType>,
    pub descendant_types: Vec<NodeType>,
    pub expected_parent_types: Vec<NodeType>,
    pub expected_child_types: Vec<NodeType>,
    pub expected_ancestor_types: Vec<NodeType>,
    pub expected_descendant_types: Vec<NodeType>,
    pub event_types: Vec<NodeType>,
    pub self_event_types: Vec<NodeType>,
    pub base_struct_type: Option<StructType>,
}
